File: scripts/eval_all.mjs
#!/usr/bin/env node
/**
 * One-shot acceptance eval for a shipped vocab:
 *   1. structural sanity (no duplicate learned rules, rank/alias integrity)
 *   2. bench-8 vs gpt-4o
 *   3. 50-probe held-out set (per-domain)
 *   4. 600-line unseen corpus eval (per-domain)
 *   5. lossless roundtrip everywhere
 *
 * Usage: node scripts/eval_all.mjs [path/to/vocab.json]   (default tokenizer/vocab.json)
 */
import { readFileSync } from 'fs';
import { encode } from '../src/encoder.js';
import { decode } from '../src/decoder.js';
import { tokenize, detokenize } from '../src/tokenizer/index.js';
import { encode as gptEncode } from 'gpt-tokenizer';
import { HELDOUT, HELDOUT_PROBES } from './heldout_probes.mjs';

const BENCH = [
  ['Common English', 'the quick brown fox jumps over the lazy dog this is a test of the emergency broadcast system how now brown cow the rain in spain stays mainly on the plain'],
  ['Code', 'const app = express(); app.get("/api/tasks", async (req, res) => { const tasks = await db.query("SELECT * FROM tasks"); res.json(tasks); });'],
  ['SQL', 'SELECT * FROM users WHERE id=42 AND name LIKE \'%test%\' ORDER BY created_at DESC; INSERT INTO table_name (a,b,c) VALUES (1,\'x\',true); UPDATE users SET name=\'abc\', score=99 WHERE id=7; DELETE FROM users WHERE id>1000; CREATE TABLE test(id INT PRIMARY KEY,name VARCHAR(255));'],
  ['API', '{"status": "success", "data": {"users": [{"id": 1, "name": "John", "email": "john@example.com"}, {"id": 2, "name": "Jane", "email": "jane@example.com"}], "total": 2, "page": 1, "per_page": 10}}'],
  ['Shell', '$ echo "Hello, World!"; printf \'%s\\n\' "$HOME"; ls -la /tmp | grep ".log" && cat file.txt; sudo -n true || echo "no sudo"; python3 -c \'print("test")\'; node -e "console.log(42)"; git status --short; git add . && git commit -m "test"; npm run build && npm start'],
  ['Markdown', '# README.md ## Test Project ### Features - fast - simple - random ### Code `npm install && npm run dev` **bold** *italic* [link](https://example.com) > quote --- ### End'],
  ['Paths', 'C:\\Users\\Test\\file.txt D:\\Games\\MC\\server.exe /usr/bin/bash ~/.config/hypr/hyprland.conf ../../src/main.js ./build/output.log https://example.com/?a=1&b=2 ftp://x@y.z:21/path git@host:user/repo.git user@example.com test+tag@example.org'],
  ['Prompt', 'aicl is Goated BTW, and this can reduce tokens very vary fast'],
];

const path = process.argv[2] ?? 'tokenizer/vocab.json';
const raw = JSON.parse(readFileSync(path, 'utf-8'));
const vocab = {
  merges: new Map(raw.merges.map(([id, rule]) => [id, rule])),
  mergeBase: raw.mergeBase ?? 100000,
  numMerges: raw.numMerges ?? raw.merges.length,
  maxTokenLength: raw.maxTokenLength ?? 5,
};

// 1. structural sanity
const seenPairs = new Set();
let dups = 0, badRanks = 0, aliasRules = 0;
for (const [id, r] of vocab.merges) {
  if (r.rank !== (id - vocab.mergeBase)) badRanks++;
  if (r.alias) { aliasRules++; continue; }
  const k = r.a + ':' + r.b;
  if (seenPairs.has(k)) dups++;
  seenPairs.add(k);
}
console.log(`vocab: ${vocab.numMerges} rules (${aliasRules} alias + ${vocab.numMerges - aliasRules} learned), maxLen=${vocab.maxTokenLength}`);
console.log(`sanity: dup learned pairs=${dups} bad ranks=${badRanks}`);
if (dups > 0) console.log('*** FAIL: duplicate learned rules — vocab is broken ***');

function evalSet(set, label) {
  let tok = 0, gpt = 0, rt = true, min = Infinity;
  for (const [, text] of set) {
    const ids = tokenize(encode(text).output, vocab);
    if (decode(detokenize(ids, vocab)).output !== text) rt = false;
    const g = gptEncode(text, { model: 'gpt-4o' }).length;
    min = Math.min(min, g / ids.length);
    tok += ids.length; gpt += g;
  }
  console.log(`${label}: AICL=${tok} gpt4o=${gpt} win=${(gpt / tok).toFixed(2)}x min=${min.toFixed(2)} lossless=${rt}`);
  return { tok, gpt, rt };
}

console.log('\n== bench-8 ==');
const b = evalSet(BENCH, 'bench ');
for (const [name, text] of BENCH) {
  const ids = tokenize(encode(text).output, vocab).length;
  const g = gptEncode(text, { model: 'gpt-4o' }).length;
  console.log(`  ${name.padEnd(15)} AICL=${String(ids).padStart(3)} gpt4o=${String(g).padStart(3)} win=${(g / ids).toFixed(2)}x`);
}

console.log('\n== held-out 50 ==');
evalSet(HELDOUT, 'held50');
for (const [domain, lines] of Object.entries(HELDOUT_PROBES)) {
  const d = evalSet(lines.map((t, i) => [`${domain}${i}`, t]), `  ${domain.padEnd(9)}`);
  if (!d.rt) console.log(`    *** ROUNDTRIP FAIL in ${domain} ***`);
}

console.log('\n== unseen corpus (600 lines, 150/domain) ==');
const unseen = JSON.parse(readFileSync(new URL('./unseen_eval.json', import.meta.url), 'utf-8'));
let allRt = true;
for (const [domain, lines] of Object.entries(unseen)) {
  const d = evalSet(lines.map((t, i) => [`${domain}${i}`, t]), `  ${domain.padEnd(9)}`);
  if (!d.rt) allRt = false;
}
console.log(`\nALL LOSSLESS: ${b.rt && allRt ? 'YES' : 'NO ***'}`);

File: scripts/eval_all.py
#!/usr/bin/env python3
"""One-shot acceptance evaluation for an AICL vocabulary.

The checks mirror the legacy acceptance script: merge-table sanity, the eight
benchmark texts, the checked-in held-out probes, and the optional unseen
corpus JSON.  It uses only the public Python runtime and the standard library.
"""

from __future__ import annotations

import argparse
import json
import sys
from collections.abc import Mapping
from pathlib import Path
from typing import Any, Sequence

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

try:
    from .eval_vocab import (  # noqa: E402
        _load_vocab,
        _tokenizer_api,
        evaluate,
        self_check as eval_vocab_self_check,
        vocab_summary,
    )
    from .heldout_probes import BENCH, HELDOUT, HELDOUT_PROBES  # noqa: E402
except ImportError:
    from eval_vocab import (  # noqa: E402
        _load_vocab,
        _tokenizer_api,
        evaluate,
        self_check as eval_vocab_self_check,
        vocab_summary,
    )
    from heldout_probes import BENCH, HELDOUT, HELDOUT_PROBES  # noqa: E402


def structural_sanity(vocab: Any) -> tuple[int, int, int]:
    merge_count, _max_length, _version = vocab_summary(vocab)
    merges = vocab.get("merges", {}) if isinstance(vocab, Mapping) else getattr(vocab, "merges", {})
    entries = merges.items() if isinstance(merges, Mapping) else merges
    seen_pairs: set[tuple[int, int]] = set()
    duplicates = 0
    bad_ranks = 0
    aliases = 0
    merge_base = int(
        vocab.get("mergeBase", vocab.get("merge_base", 100000))
        if isinstance(vocab, Mapping)
        else getattr(vocab, "merge_base", getattr(vocab, "mergeBase", 100000))
    )
    for token_id, rule in entries:
        if isinstance(rule, Mapping):
            if rule.get("alias") is True:
                aliases += 1
                continue
            rank = rule.get("rank")
            pair = (int(rule["a"]), int(rule["b"]))
        else:
            rank = rule[2] if len(rule) > 2 else None
            pair = (int(rule[0]), int(rule[1]))
        if rank != int(token_id) - merge_base:
            bad_ranks += 1
        if pair in seen_pairs:
            duplicates += 1
        seen_pairs.add(pair)
    return duplicates, bad_ranks, aliases


def _unseen(path: Path) -> dict[str, list[str]]:
    raw = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(raw, dict):
        raise ValueError("unseen evaluation file must contain an object")
    return {str(domain): [str(line) for line in lines] for domain, lines in raw.items()}


def self_check() -> None:
    eval_vocab_self_check()
    print("eval_all self-check: acceptance inputs ok")


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("vocab", nargs="?", type=Path, default=ROOT / "tokenizer" / "vocab.json")
    parser.add_argument("--unseen", type=Path, default=Path(__file__).with_name("unseen_eval.json"))
    parser.add_argument("--skip-unseen", action="store_true")
    parser.add_argument("--strict", action="store_true")
    parser.add_argument("--self-check", action="store_true")
    args = parser.parse_args(argv)
    if args.self_check:
        self_check()
        return 0

    from aicl import decode

    _load, tokenize, detokenize = _tokenizer_api()
    vocab = _load_vocab(args.vocab)
    merge_count, max_length, _version = vocab_summary(vocab)
    duplicates, bad_ranks, aliases = structural_sanity(vocab)
    print(f"vocab: {merge_count} rules ({aliases} alias + {merge_count - aliases} learned), maxLen={max_length}")
    print(f"sanity: duplicate learned pairs={duplicates} bad ranks={bad_ranks}")
    if duplicates or bad_ranks:
        print("*** FAIL: vocabulary merge table is structurally invalid ***")

    bench = evaluate(BENCH, vocab, tokenize=tokenize, detokenize=detokenize, decode=decode)
    print("\n== bench-8 ==")
    print(f"bench : AICL={bench['tokens']} tokens, CPT={bench['cpt']:.2f}, lossless={bench['lossless']}")
    for row in bench["rows"]:
        print(f"  {row['name']:<15} AICL={row['tokens']:>3} CPT={row['cpt']:.2f} rt={'ok' if row['roundtrip'] else 'FAIL'}")

    held = evaluate(HELDOUT, vocab, tokenize=tokenize, detokenize=detokenize, decode=decode)
    print("\n== held-out probes ==")
    print(f"held55: AICL={held['tokens']} tokens, CPT={held['cpt']:.2f}, lossless={held['lossless']}")
    for domain, lines in HELDOUT_PROBES.items():
        result = evaluate(
            [(f"{domain}{index}", text) for index, text in enumerate(lines)],
            vocab,
            tokenize=tokenize,
            detokenize=detokenize,
            decode=decode,
        )
        print(f"  {domain:<9} AICL={result['tokens']:>4} CPT={result['cpt']:.2f} lossless={result['lossless']}")

    all_lossless = bench["lossless"] and held["lossless"]
    if not args.skip_unseen:
        print("\n== unseen corpus ==")
        unseen = _unseen(args.unseen)
        for domain, lines in unseen.items():
            result = evaluate(
                [(f"{domain}{index}", text) for index, text in enumerate(lines)],
                vocab,
                tokenize=tokenize,
                detokenize=detokenize,
                decode=decode,
            )
            all_lossless = all_lossless and result["lossless"]
            print(
                f"  {domain:<9} lines={len(lines):>3} AICL={result['tokens']:>5} "
                f"CPT={result['cpt']:.2f} lossless={result['lossless']}"
            )
    print(f"\nALL LOSSLESS: {'YES' if all_lossless else 'NO ***'}")
    return 1 if args.strict and (duplicates or bad_ranks or not all_lossless) else 0


if __name__ == "__main__":
    raise SystemExit(main())
